### PR TITLE
topology2: cavs-nocodec: Add an intermediate pipeline

### PR DESCRIPTION
**Pull request by ranj063**
Add an intermediate pipeline between the host and DAI pipelines.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

**Review by plbossart — APPROVED** [review]
This is good to test the intermediate pipeline triggers but it should also be updated with the two-stage mixer if and when we figure out what to do with sink_c->sink and source_c->source

**Comment by aiChaoSONG** [comment]
does the mixout->gain->mixin pipeline have to be present only in playback? I think it is also possible in a capture pipeline.

**Comment by ranj063** [comment]
> does the mixout->gain->mixin pipeline have to be present only in playback? I think it is also possible in a capture pipeline.

maybe but I guess for a test for intermediate pipelines, this is good enough

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-mixin-playback.conf [hunk 1]
```diff
@@ -0,0 +1,95 @@
+#
+# BE playback pipeline: mixout-gain-mixin
+#
+# All attributes defined herein are namespaced
+# by alsatplg to "Object.Pipeline.mixout-gain-mixin-playback.N.attribute_name"
+#
+# Usage: mixout-gain-dai-copier-playback pipeline object can be instantiated as:
```
Comment (RanderWang) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-mixin-playback.conf:7]: mixout-gain-mixin-playback ?

**Review by RanderWang — APPROVED** [review]

**Review by aiChaoSONG — APPROVED** [review]

**Review by kv2019i — DISMISSED** [review]
Almost 100% failure rate in Intel IPC4 plan, see test run 16208 . UPDATE: this can be unrelated as I see FW load failing, but needs to be confirmed.

**Comment by miRoox** [comment]
SOFCI TEST

**Comment by aiChaoSONG** [comment]
> Almost 100% failure rate in Intel IPC4 plan, see test run 16208 . UPDATE: this can be unrelated as I see FW load failing, but needs to be confirmed.

I guess it is because this PR should be tested with the intermediate pipeline kernel support https://github.com/thesofproject/linux/pull/3917

**Comment by lgirdwood** [comment]
> > Almost 100% failure rate in Intel IPC4 plan, see test run 16208 . UPDATE: this can be unrelated as I see FW load failing, but needs to be confirmed.
> 
> I guess it is because this PR should be tested with the intermediate pipeline kernel support [thesofproject/linux#3917](https://github.com/thesofproject/linux/pull/3917)

@aiChaoSONG @kv2019i any update ? Can we merge ?

**Review by lgirdwood — APPROVED** [review]
@ranj063 looks like we have a conflict now.

**Comment by ranj063** [comment]
> @ranj063 looks like we have a conflict now.

@lgirdwood this will have to wait until we bring back the mixer to the nocodec tplg. We're very close to doing that. Please hold off on merging this for a couple more days

**Comment by lgirdwood** [comment]
> > @ranj063 looks like we have a conflict now.
> 
> @lgirdwood this will have to wait until we bring back the mixer to the nocodec tplg. We're very close to doing that. Please hold off on merging this for a couple more days

ok, pls ping me when ready.

**Comment by lyakh** [comment]
how is this related to #6131 ?

**Comment by plbossart** [comment]
the goal is the same @lyakh, for both topology1 and 2 we want to catch invalid firmware assumptions that a pipeline is connected to a dai or a host, and in the case of IPC4 detect cases where the intermediate pipeline is not correctly configured by the Linux driver.
Think of the nocodec topology as a 'pipe cleaner'

**Comment by kv2019i** [comment]
@ranj063  array-support merged in #7113 so rebase will be needed

**Comment by lyakh** [comment]
> @ranj063 array-support merged in #7113 so rebase will be needed

@kv2019i wouldn't this be dropped in favour of #6981 or do we need both stream types in nocodec?

**Pull request closed without merging**